crates/key-value-azure/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+
+                let cosmos_client = CosmosClient::new(self.account.clone(), token);
+                let database_client = cosmos_client.database_client(self.database.clone());
+
+                let collection_client = database_client.collection_client(self.collection.clone());
+                Ok(Arc::new(Mutex::new(collection_client)))
+            })
+        })
+        .await?;
+
+        Ok(Arc::new(AzureCosmosStore {
+            _name: name.to_owned(),
+            client: client.clone(),
+        }))
+    }
+}
+
+struct AzureCosmosStore {
+    _name: String,
+    client: Arc<Mutex<CollectionClient>>,
+}
+
+#[async_trait]
+impl Store for AzureCosmosStore {
+    async fn get(&self, key: &str) -> Result<Vec<u8>, Error> {
+        let pair = self.get_pair(key).await?;
+        Ok(pair.value)
+    }
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error> {
+        let pair = Pair {
+            id: key.to_string(),
+            value: value.to_vec(),
+        };
+        self.client
+            .lock()
+            .await
+            .create_document(pair)
+            .is_upsert(true)
+            .await
+            .map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Error> {
+        let pair = self.get_pair(key).await?;
+        let document_client = self
+            .client
+            .lock()
+            .await
+            .document_client(&pair.id, &pair.id)
+            .map_err(log_error)?;
+        document_client.delete_document().await.map_err(log_error)?;
+        Ok(())
+    }
+
+    async fn exists(&self, key: &str) -> Result<bool, Error> {
+        Ok(self.get_keys().await?.contains(&key.to_string()))
+    }
+
+    async fn get_keys(&self) -> Result<Vec<String>, Error> {
+        self.get_keys().await
+    }
+}
+
+impl AzureCosmosStore {
+    async fn get_pair(&self, key: &str) -> Result<Pair, Error> {
+        let client = self.client.lock().await;
+        let query = client
+            .query_documents(Query::new(format!("SELECT * FROM c WHERE c.id='{}'", key)))
+            .query_cross_partition(true)
+            .max_item_count(1);
+
+        // There can be no duplicated keys, so we create the stream and only take the first result.
+        let mut stream = query.into_stream::<Pair>();
+        let res = stream.next().await;
+        match res {
+            Some(r) => {
+                let r = r.map_err(log_error)?;
+                match r.results.first() {
+                    Some(p) => Ok(p.0.clone()),
+                    None => Err(Error::NoSuchKey),
+                }
+            }
+            None => Err(Error::NoSuchKey),
+        }
+    }
+
+    async fn get_keys(&self) -> Result<Vec<String>, Error> {
+        let client = self.client.lock().await;
+        let query = client
+            .query_documents(Query::new("SELECT * FROM c".to_string()))
```

Comment:
I _think_ we can tweak the query to only return `c.id` instead of the entire document?